### PR TITLE
Updates to doc structures and using rustdoc format

### PR DESCRIPTION
**Pull request by yashaswi2000**
## Description

Reformatting documentation to follow rustdoc format. enabling us to move towards Rust crate style documentation, supporting auto doc generation. 
will be adding general templates for documenting structs, functions, traits here - [style guide](https://github.com/Lind-Project/lind-docs/blob/main/docs/RustPOSIX/Style-Guide.md)

Following rustdoc book for overall guidance and rust docs conventions from microsoft opensource resources.
[rustdoc book](https://doc.rust-lang.org/rustdoc/what-is-rustdoc.html)
[templates](https://microsoft.github.io/mu/CodeDevelopment/rust_documentation_conventions/#crate-template)

Fixes # (issue)

 

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

- Test A - `lind_project/tests/test_cases/test_a.c`
- Test B - `lind_project/tests/test_cases/test_b.c`

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to a pull request and/or merged in other modules (native-client, lind-glibc, lind-project)

### CODE REVIEW

File: src/safeposix/syscalls/fs_calls.rs
Comment (JustinCappos) [inline src/safeposix/syscalls/fs_calls.rs]: This may cause problems with the test runner.  It assumes that more than 4 spaces is code to execute.

File: src/safeposix/syscalls/fs_calls.rs
Comment (JustinCappos) [inline src/safeposix/syscalls/fs_calls.rs]: Text with _ or other special characters is often treated as markdown by rustdoc.   I think using backquotes around function / variable names fixes this...

File: src/safeposix/syscalls/fs_calls.rs
Comment (JustinCappos) [inline src/safeposix/syscalls/fs_calls.rs]: What is this?

**Review by JustinCappos — CHANGES_REQUESTED** [review]
There are a few minor lessons learned from my work with rustdoc on fdtables.  Let me know if any of my feedback isn't clear...

File: src/safeposix/syscalls/fs_calls.rs
Comment (JustinCappos) [inline src/safeposix/syscalls/fs_calls.rs:2053]: I think it's typical to have a 'panic' block as well, which lists all the instances in which the code panics.

File: src/safeposix/syscalls/fs_calls.rs
Comment (JustinCappos) [inline src/safeposix/syscalls/fs_calls.rs:295]: I don't think we need this section.  They can look at the tests and find situations where there is or is not a test.

File: src/safeposix/syscalls/fs_calls.rs
Comment (JustinCappos) [inline src/safeposix/syscalls/fs_calls.rs]: This should list all of the error codes and exactly when they happen.

File: src/safeposix/syscalls/fs_calls.rs
Comment (JustinCappos) [inline src/safeposix/syscalls/fs_calls.rs:15]: Note, that Rustdoc will automatically create a list like this for you for each of the functions.  I'm not sure you need to do this manually.

File: src/lib.rs
Comment (JustinCappos) [inline src/lib.rs]: These sorts of things are usually not comments for Rustdoc (which is meant to be read by users of the library, not developers working on it.)

File: src/interface/mod.rs
Comment (JustinCappos) [inline src/interface/mod.rs]: Do we hide this or expose it to people using it?  If we hide it, we don't need to explain this structure in rustdoc.  It could be a normal comment.

In other words, if I need to know this to write a program that uses this module, you should tell me.  If not, then you can hide this detail.

File: src/safeposix/cage.rs
Comment (JustinCappos) [inline src/safeposix/cage.rs:19]: Are these sorts of things exposed to the user?

**Review by JustinCappos — COMMENTED** [review]
A bunch of comments here about what does and doesn't belong in Rustdoc.  I'm happy to discuss further.

File: src/safeposix/syscalls/fs_calls.rs
Comment (rennergade) [reply src/safeposix/syscalls/fs_calls.rs]: This is reflected in the errnos file and perhaps should just link to there? It feels redundant to list every errno at the top of every file.

Comment (rennergade) [reply src/safeposix/syscalls/fs_calls.rs]: oh I misread, this just lists for per syscall, that make sense to me

File: src/safeposix/syscalls/fs_calls.rs
Comment (yashaswi2000) [reply src/safeposix/syscalls/fs_calls.rs:2053]: will update it.

File: src/safeposix/syscalls/fs_calls.rs
Comment (yashaswi2000) [reply src/safeposix/syscalls/fs_calls.rs:295]: will link the tests doc here for the functions

File: src/safeposix/syscalls/fs_calls.rs
Comment (yashaswi2000) [reply src/safeposix/syscalls/fs_calls.rs:15]: So, the sys calls are implemented as part of Cage struct. and we implemented them in different modules for better maintainability. but rustdoc associates these functions to Cage struct and these modules turn up empty. so added these lists to redirect to the Cage docs.

File: src/lib.rs
Comment (yashaswi2000) [reply src/lib.rs]: will refactor

File: src/safeposix/cage.rs
Comment (yashaswi2000) [reply src/safeposix/cage.rs:19]: we are exposing these structs as public through File descriptor enums

File: src/interface/mod.rs
Comment (yashaswi2000) [reply src/interface/mod.rs]: understood. will refactor

File: src/safeposix/dispatcher.rs
Comment (rennergade) [inline src/safeposix/dispatcher.rs]: I don't think this is true for the dispatcher? What are you referring to here?

File: src/safeposix/mod.rs
Comment (rennergade) [inline src/safeposix/mod.rs]: I'm not sure you need this first sentence. Most people won't know what this means.

Comment (yashaswi2000) [reply src/safeposix/mod.rs]: will refactor

**Review by rennergade — CHANGES_REQUESTED** [review]
A few minor change requests but mostly this looks good to go.

File: src/safeposix/dispatcher.rs
Comment (yashaswi2000) [reply src/safeposix/dispatcher.rs]: was trying to mention about the lock wrapping we do around objects, but that seems to not very relevant here. clearing that up

**Review by rennergade — APPROVED** [review]
Approved!

**Comment by rennergade** [comment]
@JustinCappos Id love to get your approval on this one too before I merge it.

File: src/safeposix/syscalls/fs_calls.rs
Comment (JustinCappos) [reply src/safeposix/syscalls/fs_calls.rs:15]: Okay, please make sure you indicate this in a comment in the code

**Review by JustinCappos — APPROVED** [review]
Looks good.  Please do explain why the functions are listed the way they are for posterity.  Otherwise, I think it's okay to merge.

File: src/safeposix/syscalls/fs_calls.rs
Comment (JustinCappos) [reply src/safeposix/syscalls/fs_calls.rs:15]: (By the way, you can look at fdtables, which has a similar issue but gets around it.  This may change the visibility / import structure of the code though.  I honestly, don't recall exactly how I did this.)

**Pull request merged**
